server/src/server/kb/concept/ValueConverter.java

```diff
@@ -0,0 +1,96 @@
+/*
+ * GRAKN.AI - THE KNOWLEDGE GRAPH
+ * Copyright (C) 2018 Grakn Labs Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as
+ * published by the Free Software Foundation, either version 3 of the
+ * License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ */
+
+package grakn.core.server.kb.concept;
+
+import com.google.common.collect.ImmutableMap;
+import grakn.core.concept.type.AttributeType;
+import java.time.Instant;
+import java.time.LocalDateTime;
+import java.time.ZoneId;
+import java.util.Map;
+
+public abstract class ValueConverter<SOURCE, TARGET>{
+
+    private static Map<AttributeType.DataType<?>, ValueConverter<?, ?>> converters = ImmutableMap.<AttributeType.DataType<?>, ValueConverter<?, ?>>builder()
+            .put(AttributeType.DataType.BOOLEAN, new IdentityConverter<Boolean, Boolean>())
+            .put(AttributeType.DataType.DATE, new DateConverter())
+            .put(AttributeType.DataType.DOUBLE, new DoubleConverter())
+            .put(AttributeType.DataType.FLOAT, new FloatConverter())
+            .put(AttributeType.DataType.INTEGER, new IntegerConverter())
+            .put(AttributeType.DataType.LONG, new LongConverter())
+            .put(AttributeType.DataType.STRING, new IdentityConverter<String, String>())
+            .build();
+
+    public static <SOURCE, TARGET> ValueConverter<SOURCE, TARGET> of(AttributeType.DataType<TARGET> dataType) {
+        ValueConverter<?, ?> converter = converters.get(dataType);
+        if (converter == null){
+            throw new UnsupportedOperationException("Unsupported DataType: " + dataType.toString());
+        }
+        return (ValueConverter<SOURCE, TARGET>) converter;
+    }
+
+    public abstract TARGET convert(SOURCE value);
+
+    public static class IdentityConverter<SOURCE, TARGET> extends ValueConverter<SOURCE, TARGET> {
+        @Override
+        public TARGET convert(SOURCE value) { return (TARGET) value;}
+    }
+
+    public static class DateConverter extends ValueConverter<Object, LocalDateTime> {
+
+        @Override
+        public LocalDateTime convert(Object value) {
+            if (value instanceof LocalDateTime) return (LocalDateTime) value;
+            else if (value instanceof Long) {
+                LocalDateTime.ofInstant(Instant.ofEpochMilli((Long) value), ZoneId.of("Z"));
```

Comment:
Shouldn't there be a `return` statement here, @kasper-piskorski ?

Reply:
good spot!